src/container/srv_container.c

```diff
@@ -371,6 +371,75 @@ ds_cont_init_metadata(struct rdb_tx *tx, const rdb_path_t *kvs,
 	return rc;
 }
 
+/* check if container exists by UUID and (if applicable) non-default label */
+static int
+cont_existence_check(struct rdb_tx *tx, struct cont_svc *svc,
+		     uuid_t puuid, uuid_t cuuid, char *clabel)
+{
+	d_iov_t		key;
+	d_iov_t		val;
+	bool		may_exist = false;
+	uuid_t		match_cuuid;
+	int		rc;
+
+	/* Find by UUID in cs_conts KVS. */
+	d_iov_set(&key, cuuid, sizeof(uuid_t));
+	d_iov_set(&val, NULL /* buf */, 0 /* size */);
+	rc = rdb_tx_lookup(tx, &svc->cs_conts, &key, &val);
+	if (rc != -DER_NONEXIST) {
+		if (rc != 0)
+			return rc;	/* other lookup failure */
+		may_exist = true;
+	}
+
+	/* If no label in request, return cs_conts lookup result */
+	if (clabel == NULL) {
+		D_DEBUG(DF_DSMS, DF_CONT": no label, lookup by UUID "DF_UUIDF
+			" "DF_RC"\n", DP_CONT(puuid, cuuid), DP_UUID(cuuid),
+			DP_RC(rc));
+		return rc;
+	}
+
+	/* Label provided in request - search for it in cs_uuids KVS
+	 * and perform additional sanity checks.
+	 */
+	d_iov_set(&key, clabel, strnlen(clabel, DAOS_PROP_LABEL_MAX_LEN+1));
+	d_iov_set(&val, match_cuuid, sizeof(uuid_t));
+	rc = rdb_tx_lookup(tx, &svc->cs_uuids, &key, &val);
+	if (rc != -DER_NONEXIST) {
+		if (rc == 0) {
+			/* not found by UUID, but label matched - invalid */
+			if (!may_exist) {
+				D_ERROR(DF_CONT": non-unique label: %s\n",
+					DP_CONT(puuid, cuuid), clabel);
+				return -DER_INVAL;
+			}
+
+			/* found by UUID, and found by label: make sure the
+			 * label lookup returned same UUID as the request UUID.
+			 */
+			if (uuid_compare(cuuid, match_cuuid)) {
+				D_ERROR(DF_CONT": label=%s -> "DF_UUID
+					" (mismatch)\n", DP_CONT(puuid, cuuid),
+					clabel, DP_UUID(match_cuuid));
+				return -DER_INVAL;
+			}
```

Comment:
I find myself wondering about all these sanity checks. Is there some case where this granularity is useful? If cont create is the only one using it, we would want this to return an error if uuid exists OR label exists, right? You may be able to simplify quite a bit and get the same behavior.

Reply:
Got it. Thanks for the context.